### PR TITLE
chore: add config and code run/monitor just-nuts @W-8969970@

### PR DESCRIPTION
**Pull request by peternhale**
### What does this PR do?
Adds automation for running NUTs across plugins

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,6 +3,14 @@ version: 2.1
 orbs:
   release-management: salesforce/npm-release-management@4
 
+parameters:
```
Comment (peternhale) [inline .circleci/config.yml:6]: Pipeline parameters used to run only NUT tests. Code that triggers/monitors will set `run-auto-workflows` to false and `run-just-nuts` to true. With all workflows guarded with one of these vars it will control which run when pipelines run, with the goal of only running NUTs tests when triggering just nuts from last step of sfdx-cli release pipeline.

File: .circleci/config.yml [hunk 2]
```diff
@@ -133,9 +141,37 @@ jobs:
       - run: yarn install
       - run: ./scripts/promote
 
+  run-all-nuts:
```
Comment (peternhale) [inline .circleci/config.yml:144]: Job to trigger NUT tests in other pipelines

File: .circleci/config.yml [hunk 3]
```diff
@@ -133,9 +141,37 @@ jobs:
...
+      - image: node:latest
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - v3-npm-{{checksum "yarn.lock"}}
+      - run: yarn install
+      - run: ./scripts/run-plugin-nuts.js
+
+  monitor-all-nuts:
```
Comment (peternhale) [inline .circleci/config.yml:155]: Job that takes in pipeline instance data and then monitors each one until completion.

File: .circleci/config.yml [hunk 4]
```diff
@@ -133,9 +141,37 @@ jobs:
...
+      - restore_cache:
+          keys:
+            - v3-npm-{{checksum "yarn.lock"}}
+      - run: yarn install
+      - run: ./scripts/monitor-plugin-nuts.js <<parameters.nuts_job_data>>
+
 workflows:
   version: 2.1
   test-and-release:
+    when: << pipeline.parameters.run-auto-workflows >>
```
Comment (peternhale) [inline .circleci/config.yml:174]: Condition that controls if a workflow will run. For typical PR/Merge activities, this workflow is triggered by GH events. The default value for `run-auto-workflows` is true, so this workflow will run.

File: .circleci/config.yml [hunk 5]
```diff
@@ -231,3 +267,19 @@ workflows:
...
               only:
                 - main
+      - run-all-nuts:
+          filters:
+            branches:
+              only:
+                - main
+
+  monitor-all-nuts:
+    when: << pipeline.parameters.run-just-nuts >>
```
Comment (peternhale) [inline .circleci/config.yml:277]: The condition that prevents this workflow from running under typical conditions, `run-just-nuts` is false by default.

File: package.json [hunk 6]
```diff
@@ -141,6 +141,7 @@
     "eslint-plugin-jsdoc": "^27.0.3",
     "eslint-plugin-prettier": "^3.1.3",
     "globby": "^11.0.2",
+    "got": "^11.8.2",
```
Comment (peternhale) [inline package.json:144]: Need a library to do REST calls, `got` seems very popular so I grabbed it. Might be a suitable replacement for `request`.

File: scripts/monitor-plugin-nuts.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env node
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:1]: node script used to monitor remote NUT tests

File: scripts/monitor-plugin-nuts.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  getCircleCiUrl() {
+    return `https://app.circleci.com/pipelines/${this.justNuts.project_slug}/${this.justNuts.pipeline_number}/workflows/${this.justNuts.id}`;
+  }
+
+  async circle(url, options = {}) {
+    const defaultOptions = {
+      method: 'GET',
+      headers: { 'Circle-Token': process.env.CIRCLE_TOKEN, responseType: 'json', resolveBodyOnly: true },
+    };
+    options = Object.assign(defaultOptions, { ...options });
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:48]: I'll change to `{...defaultOptions, ...options }`

File: scripts/monitor-plugin-nuts.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    try {
+      return JSON.parse((await got(url, options)).body);
+    } catch (error) {
+      console.log(error);
+      throw new Error(error && error.response?.body ? error.response.body : error);
+    }
+  }
+
+  async checkWorkflowState() {
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:57]: Checks workflow state. Right now this is only interested in workflows with the name "just-nuts"

File: scripts/monitor-plugin-nuts.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    this.isComplete = notCompleteStatus.some((status) => justNuts.status !== status);
+    this.justNuts = justNuts;
+  }
+
+  async waitForCompletion() {
+    const retries = 30;
+    let retryCnt = 0;
+    while (!this.isComplete && retryCnt++ <= retries) {
+      await sleep(Duration.seconds(5));
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:73]: I set a time limit of 15 minutes to wait for a job to complete
duration needs to be 30. I'll fix.

File: scripts/monitor-plugin-nuts.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    if (retryCnt > retries && !this.isComplete) {
+      this.status = 'monitor timed out';
+    }
+  }
+  displayWorkflowState() {
+    console.log(`Workflow: ${this.justNuts.name} Status: ${this.justNuts.status} URL: ${this.getCircleCiUrl()}`);
+  }
+}
+
+if (process.argv.length < 3) {
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:85]: data can only be passed as a string, expecting a JSON string.

File: scripts/monitor-plugin-nuts.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+  throw Error('did not receive expected input');
+}
+
+let args;
+try {
+  args = JSON.parse(process.argv[2]);
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:91]: data can only be passed as a string, expecting a JSON string.

File: scripts/monitor-plugin-nuts.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+} catch (error) {
+  throw new Error(`error parsing json input ${process.argv[2]}`);
+}
+
+if (!isArray(args)) {
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:96]: JSON string show represent an array

File: scripts/monitor-plugin-nuts.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+  throw new Error('expecting input to be a JSON array');
+}
+
+args = [
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:100]: Cruft - need to delete

File: scripts/monitor-plugin-nuts.js [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    org: 'salesforcecli',
+    repo: 'data',
+    id: 'c1be631e-32ba-4110-bbd5-43d56b50e33a',
+    state: 'created',
+    number: 0,
+    created_at: '2019-08-24T14:15:22Z',
+  },
+];
+
+const monitors = args.map((arg) => new MonitorPluginNuts(arg));
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:111]: Instantiate monitor for each job

File: scripts/monitor-plugin-nuts.js [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    const justNuts = response.items.find((item) => item.name === 'just-nuts');
+    if (!justNuts) {
+      this.isComplete = true;
+      return;
+    }
+    this.isComplete = notCompleteStatus.some((status) => justNuts.status !== status);
+    this.justNuts = justNuts;
+  }
+
+  async waitForCompletion() {
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:69]: Wait for completion

File: scripts/monitor-plugin-nuts.js [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    number: 0,
+    created_at: '2019-08-24T14:15:22Z',
+  },
+];
+
+const monitors = args.map((arg) => new MonitorPluginNuts(arg));
+
+let rc = 0;
+
+(async () => {
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:115]: Main logic of script

File: scripts/monitor-plugin-nuts.js [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+  console.log(`Begin monitoring of ${monitors.length} workflows`);
+  await Promise.all(monitors.map(async (monitor) => monitor.waitForCompletion()));
+  const exitCode = monitors.some((monitor) => monitor.justNuts?.status !== 'success') ? 1 : 0;
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:118]: establish exit code

File: scripts/monitor-plugin-nuts.js [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+  console.log(
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:119]: Summary of results to console

File: scripts/monitor-plugin-nuts.js [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    `End monitoring. Successful workflows ${
+      monitors.filter((monitor) => monitor.status === 'success').length
+    }. Unsuccessful workflows ${monitors.filter((monitor) => monitor.status !== 'success').length}`
+  );
+  monitors.forEach((monitor) => monitor.displayWorkflowState());
```
Comment (peternhale) [inline scripts/monitor-plugin-nuts.js:124]: Display individual workflow results

File: scripts/run-plugin-nuts.js [hunk 21]
```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env node
```
Comment (peternhale) [inline scripts/run-plugin-nuts.js:1]: node script to trigger NUTs for other pipelines

File: scripts/run-plugin-nuts.js [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+      .reduce((a, b) => {
+        return a.concat(b);
+      }, [])
+      // find plugins that have a script named 'test:nuts'
+      .filter((plugin) => hasTestNuts(plugin))
+      // establish github org and repo name
+      .map((plugin) => Object.assign({}, { plugin, info: getOrgAndProject(plugin) }))
+  );
+};
+
```
Comment (peternhale) [inline scripts/run-plugin-nuts.js:138]: Need to add logic to process input arguments

File: scripts/monitor-plugin-nuts.js (hunk 10 above)
Comment (peternhale) [reply scripts/monitor-plugin-nuts.js:73]: Should retries or sleep duration be exposed as env vars or parameters?

File: scripts/monitor-plugin-nuts.js [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+
+if (!isArray(args)) {
+  throw new Error('expecting input to be a JSON array');
+}
+
+const monitors = args.map((arg) => new MonitorPluginNuts(arg));
+
+(async () => {
+  console.log(`Begin monitoring of ${monitors.length} workflows`);
+  await Promise.all(monitors.map(async (monitor) => monitor.waitForCompletion()));
```
Comment (amphro) [inline scripts/monitor-plugin-nuts.js:129]: Nice!

File: scripts/monitor-plugin-nuts.js [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+      console.log(error);
+      throw new Error(error && error.response?.body ? error.response.body : error);
+    }
+  }
+
+  async checkWorkflowState() {
+    const url = this.getWorkflowUrl();
+    const response = await this.circle(url, {});
+    const justNuts = response.items.find((item) => item.name === 'just-nuts');
+    if (!justNuts) {
```
Comment (amphro) [inline scripts/monitor-plugin-nuts.js:86]: So this is essentially saying if there is not a job just-nut workflow, then ignore it?

File: scripts/run-plugin-nuts.js [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+/**
+ * Triggers the just-nuts workflow for the named org/repo/branch
+ * @param sfdxVersion
+ * @param org
+ * @param repo
+ * @param branch
+ * @returns {Promise<*>}
+ */
+const triggerNutsForProject = async (sfdxVersion, org, repo, branch = 'main') => {
```
Comment (amphro) [inline scripts/run-plugin-nuts.js:68]: nice

File: scripts/run-plugin-nuts.js [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+    json: body,
+  });
+};
+/**
+ * Makes a request to named URL. Ensures that auth token is applied.
+ * @param url
+ * @param options
+ * @returns {Promise<any>}
+ */
+const circle = async (url, options = {}) => {
```
Comment (amphro) [inline scripts/run-plugin-nuts.js:106]: If these scripts grow anymore, we should probably break out similar functionality and require it in.

File: .circleci/config.yml [hunk 27]
```diff
@@ -133,9 +141,43 @@ jobs:
...
+        default: 'latest-rc'
+        type: string
+
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - v3-npm-{{checksum "yarn.lock"}}
+      - run: yarn install
+      - run: ./scripts/run-plugin-nuts.js
```
Comment (peternhale) [inline .circleci/config.yml:159]: Need to pass in sfdx_version

File: scripts/run-plugin-nuts.js [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+/**
+ * Examines a lit of plugins to determine which have NUT tests
+ * @param plugins
+ * @returns {*}
+ */
+const qualifyPluginsWithNonUnitTests = (plugins) => {
+  return (
+    plugins
+      // find plugins that may have NUT tests
+      .filter((plugin) => plugin === 'salesforcedx' || plugin.startsWith('@salesforce'))
```
Comment (amphro) [inline scripts/run-plugin-nuts.js:132]: Do we really need this check? What is the harm of checking every plugin?

File: package.json (hunk 6 above)
Comment (peternhale) [reply package.json:144]: I just saw that someone pulled in `got` to plugin-release-management, so not a new discovery ;).

File: scripts/run-plugin-nuts.js [hunk 29]
```diff
@@ -0,0 +1,183 @@
...
+      // handle sub-aggregate plugins by fetching that plugins dependencies
+      .map((plugin) => {
+        if (plugin === 'salesforcedx') {
```
Comment (amphro) [inline scripts/run-plugin-nuts.js:135]: You could probably also make this generic and even recursive if you wanted. If `npm view <pkg> oclif.plugins` then also check those. I think it is fine to also keep it simple for now and make it more generic later if needed.

File: scripts/run-plugin-nuts.js [hunk 30]
```diff
@@ -0,0 +1,183 @@
...
+        } else {
+          return plugin;
+        }
+      })
+      // flatten arrays
+      .reduce((a, b) => {
+        return a.concat(b);
+      }, [])
+      // find plugins that have a script named 'test:nuts'
+      .filter((plugin) => hasTestNuts(plugin))
```
Comment (amphro) [inline scripts/run-plugin-nuts.js:147]: This is a super edge case. We run these nuts on the `latest-rc`. If a plugin published a `latest-rc` tag which introduced test:nuts, this wouldn't pick it up because it is checking that script on the latest tag.

I don't think we really need to account for this since it will rarely happen and it is only for the first week they introduce it, all other weeks will get picked up. SO I think it is fine.

**Review by amphro — APPROVED** [review]
Nice work on this.

File: scripts/run-plugin-nuts.js (hunk 30 above)
Comment (peternhale) [reply scripts/run-plugin-nuts.js:147]: @amphro  Since the goal of just-nuts is to force plugin NUT tests to reply on the installed `sfdx`, rather than `bin/run`, so I should not depend on tags applied to plugins, since those advance at a different rate than sfdx-cli. In reality, the should be checking the absolute version of plugins installed with sfdx-cli to determine if test:nuts are available and possibly change the check for plugins to checkout source for the same version found in sfdx-cli installation, just to make sure the tests run aligned with sfdx-cli version.

Comment (amphro) [reply scripts/run-plugin-nuts.js:147]: Checking the absolute version and running test from the commit has of that version would definitely be better, but maybe lets do that in a follow up? Unless you think it would be a pretty easy change to do that. It still feels like an edge case so I don't want this story to keep getting bigger.

File: scripts/run-plugin-nuts.js [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+  })();
+  return JSON.parse(response.body);
+};
+
+/**
+ * Examines a lit of plugins to determine which have NUT tests
+ * @param plugins
+ * @returns {*}
+ */
+const qualifyPluginsWithNonUnitTests = (modules) => {
```
Comment (peternhale) [inline scripts/run-plugin-nuts.js:128]: Reworked this function to take into account the version of a module when fetching details from nom,

`npm show sfdx-cli@latest-rc....` 

Npm is nice enough to returned the requested details related to the named version in the show command.

File: scripts/run-plugin-nuts.js [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+    repo: 'toolbelt',
+  },
+};
+
+const hasTestNuts = (module) => {
+  const results = execSync(`npm show ${module.name}@${module.version} scripts`);
+  return results.includes('test:nuts');
+};
+
+const getNpmForModule = (timeCreated, module, sections) => {
```
Comment (peternhale) [inline scripts/run-plugin-nuts.js:39]: @amphro I modified this function with the goal of either returning a single version of a module or undefined. sfdx-cli has a reference to plugin-auth@^1.3.0 and since that allows version selection for all version in the 1.x lineage, the plugin was always being selected as having `test:nuts`. This change takes some additional information, timeCreated, which is the timestamp assigned to the sfdx-cli version. If the npm show returns more than one version (as it does for auth) the code will sort the returned version is version timestamp descending and then locate the version of a plugin whose version timestamp is <= sfdx-cli timestamp.

This also has the benefit of fixing an exact version for any plugin selected, which will be used by NUTS to checkout the correct version of the plugin that was used at the time sfdx-cli was published.

File: scripts/run-plugin-nuts.js [hunk 33]
```diff
@@ -0,0 +1,201 @@
...
+ * @returns {Promise<*>}
+ */
+const triggerNutsForProject = async (sfdxVersion, module, branch = 'main') => {
+  const body = {
+    branch: branch,
+    parameters: {
+      'run-auto-workflows': false,
+      'run-just-nuts': true,
+      sfdx_version: sfdxVersion,
+      repo_tag: module.version,
```
Comment (peternhale) [inline scripts/run-plugin-nuts.js:82]: Passes the module.version found for a plugin as the repo_tag that will be used in checkout.

**Review by RodEsp — APPROVED** [review]
Nice! Looks good.

**Comment by RodEsp** [comment]
Is it possible to add testing steps to the description?

File: oclif.manifest.json [hunk 34]
```diff
@@ -0,0 +1 @@
+{"version":"7.97.1","commands":{}}
```
Comment (amphro) [inline oclif.manifest.json:1]: This should not be checked in.

Comment (peternhale) [reply oclif.manifest.json:1]: I'll revert the change

**Comment by peternhale** [comment]
> Is it possible to add testing steps to the description?

@RodEsp Going to need to test this during a release.

File: scripts/monitor-plugin-nuts.js (hunk 8 above)
Comment (mshanemc) [reply scripts/monitor-plugin-nuts.js:73]: any reason not to?

File: scripts/monitor-plugin-nuts.js (hunk 23 above)
Comment (mshanemc) [reply scripts/monitor-plugin-nuts.js:129]: Putting on my ops hat...I'm imagining this taking a while.  And no output until they all return?  

It might be nice to know when one repo starts/completes (console.log from the waitForCompletion or the check function it calls) so you can see what we're waiting on.

File: scripts/run-plugin-nuts.js [hunk 35]
```diff
@@ -0,0 +1,206 @@
...
+    }
+  })();
+  return JSON.parse(response.body);
+};
+
+/**
+ * Examines a lit of plugins to determine which have NUT tests
+ * @returns {*}
+ */
+const qualifyPluginsWithNonUnitTests = (timeCreated, modules) => {
```
Comment (mshanemc) [inline scripts/run-plugin-nuts.js:139]: I may not understand what this is trying to do.  Is there a reason not to look into the yarn.lock file to see what sfdx-cli resolved the plugin versions to be (since that's what it's going to build/ship) rather than working with npm, timestamps, etc?

Comment (mshanemc) [reply scripts/run-plugin-nuts.js:139]: you could probably do the same `.filter((module) => /^@*salesforce/.test(module.name));` on that?

**Review by mshanemc — APPROVED** [review]
Suggestions only.  If this is running all the NUTs, it's awesome!

File: scripts/run-plugin-nuts.js [hunk 36]
```diff
@@ -0,0 +1,206 @@
...
+          .filter((module) => /^@*salesforce/.test(module.name));
+        // determine if current module should be include (has a nut test) and qualify all of current modules dependencies
+        return [
+          ...(hasTestNuts(module) ? [module] : []),
+          ...qualifyPluginsWithNonUnitTests(timeCreated, pluginsToQualify),
+        ];
+      })
+      .filter((module) => module)
+      // flatten arrays
+      .reduce((a, b) => {
```
Comment (amphro) [inline scripts/run-plugin-nuts.js:164]: Isn't there a `.flat()` method on array now?

File: scripts/run-plugin-nuts.js (hunk 32 above)
Comment (amphro) [reply scripts/run-plugin-nuts.js:40]: Can you add a comment explaining this. Took me a second to wrap my head around this.

Also, I don't think **ANY** core plugin should be carroted, including plugin-auth. We should probably keep this code in here in case it does happen, we can handle it, but I think we should add it to the pinned dep list.

File: scripts/run-plugin-nuts.js (hunk 35 above)
Comment (amphro) [reply scripts/run-plugin-nuts.js:139]: The yarn.lock is an interesting idea. I like that better than the timestamp approach. Would that work @peternhale ?

**Review by amphro — APPROVED** [review]
